Action delete API rework - support for action file deletion

mahesh-orch opened the pull request:
* Action delete API modification to have option to delete action database entry without removing files from disk
* From CLI `st2 action delete <pack>.<action>` will delete database entry only
* From CLI `st2 action delete --remove-files <pack>.<action>` will delete database entry and files.
* If action DELETE method has additional argument as `{"remove_files": true}` in the body then it will remove related action files from disk along with database entry.
* `{"remove_files": false}` or no additional argument in action DELETE method body will remove only database entry.

m4dcoder commented on st2api/st2api/controllers/v1/actions.py:
```diff
@@ -206,7 +206,7 @@ def put(self, action, ref_or_id, requester_user):
         return action_api
 
-    def delete(self, ref_or_id, requester_user):
+    def delete(self, files_remove_request, ref_or_id, requester_user):
```
Can you rename `files_remove_request` to `options`?

m4dcoder commented on st2common/st2common/openapi.yaml.j2:
```diff
@@ -409,6 +409,11 @@ paths:
           description: Entity reference or id
           type: string
           required: true
+        - name: files_remove_request
```
Can you rename `files_remove_request` to `options`?

m4dcoder commented on st2common/st2common/openapi.yaml:
```diff
@@ -413,6 +413,11 @@ paths:
           description: Entity reference or id
           type: string
           required: true
+        - name: files_remove_request
```
Can you rename `files_remove_request` to `options`?

m4dcoder commented on st2client/st2client/commands/action.py:
```diff
@@ -282,31 +282,24 @@ def __init__(self, resource, *args, **kwargs):
-            dest="force",
-            help="Auto yes flag to delete action files from disk.",
+            dest="remove_files",
+            help="Remove action files from disk.",
```
Do we need to add `default=False`?

m4dcoder commented on st2client/st2client/commands/action.py:
```diff
@@ -282,31 +282,24 @@ def __init__(self, resource, *args, **kwargs):
+            remove_files = True
         else:
-            print("Action is not deleted.")
+            remove_files = False
```
Since the remove_files arg is already defined as boolean above, you can change this if/else block to `remove_files = args.remove_files`.

m4dcoder commented on st2api/tests/unit/controllers/v1/test_actions.py:
```diff
@@ -854,3 +856,12 @@ def __do_delete(self, action_id, expect_errors=False):
         return self.app.delete(
             "/v1/actions/%s" % action_id, expect_errors=expect_errors
         )
+
```
- Update existing unit test `test_delete` where there's no payload and assert database record is removed and mocked delete_action_files_from_pack is not called
- Add new unit test `test_delete_remove_files_false` where remove_files=False in payload and assert database record is removed and mocked delete_action_files_from_pack is not called
- Add new unit test `test_delete_remove_files_true` where remove_files=True in payload and assert database record is removed and mocked delete_action_files_from_pack is called

ashwini-orchestral commented:
@m4dcoder I have added unit tests, updated existing unit test and done other minor changes as suggested. Please review. Thanks.

mahesh-orch commented:
@m4dcoder I have added unit tests, updated existing unit test and done other minor changes as suggested. Please review. Thanks.

m4dcoder reviewed (approved):
LGTM

nzlosh commented on st2client/st2client/commands/action.py:
```diff
@@ -282,31 +282,21 @@ def __init__(self, resource, *args, **kwargs):
         self.parser.add_argument(
-            "-f",
-            "--force",
+            "-rf",
```
Single hyphen flags should be a single letter only.  Under Linux, it's possible to have flags grouped by a single hyphen which would make the `-rf` flag seem like two flags `-r -f` were being passed. `-r` should be sufficient for the `remove_files` flag unless it's already taken.

amanda11 commented on st2common/st2common/openapi.yaml.j2:
```diff
@@ -5380,6 +5388,14 @@ definitions:
+    properties:
+      remove_files:
+        type: boolean
+        description: Force delete action files from disk
```
Minor comment - should this be reworded as in ohter schema, as its not really a force - its just a flag.

amanda11 reviewed (commented):
Just a minor comment on wording in the schema, but @nzlosh point more important.

mahesh-orch commented:
@nzlosh I have changed CLI action delete optional argument `-rf` to `-r`.
@amanda11 I have updated the description of ActionDeleteSchema `remove-files` argument.
Please review. Thanks.
